File: wasm/src/dbc/values.zig
//! DBC value metadata parsing.
//!
//! Handles `VAL_`, `VAL_TABLE_`, and `SIG_VALTYPE_` records that decorate
//! signals after message parsing.

const std = @import("std");
const quotes = @import("quotes.zig");

/// Largest exact integer representable by JavaScript `number`.
const JS_SAFE_INTEGER_MAX: i64 = 9_007_199_254_740_991;

/// Numeric representation requested by `SIG_VALTYPE_`.
pub const ValueType = enum { integer, float32, float64 };

/// One raw numeric value and its display label.
pub const ValueDescription = struct { raw_value: i64, label: []const u8 };

/// Duplicates value descriptions so a signal can own a copy of a named table.
pub fn dupeValueDescriptions(allocator: std.mem.Allocator, descriptions: []const ValueDescription) ![]ValueDescription {
    const copy = try allocator.alloc(ValueDescription, descriptions.len);
    errdefer allocator.free(copy);

    var copied_labels: usize = 0;
    errdefer {
        for (copy[0..copied_labels]) |description| {
            allocator.free(description.label);
        }
    }

    for (descriptions, copy) |description, *target| {
        target.* = .{
            .raw_value = description.raw_value,
            .label = try allocator.dupe(u8, description.label),
        };
        copied_labels += 1;
    }

    return copy;
}

/// Releases only labels inside a value-description slice.
pub fn freeValueDescriptionLabels(allocator: std.mem.Allocator, descriptions: []const ValueDescription) void {
    for (descriptions) |description| {
        allocator.free(description.label);
    }
}

/// Releases labels and the slice allocated for value descriptions.
pub fn freeValueDescriptions(allocator: std.mem.Allocator, descriptions: []const ValueDescription) void {
    freeValueDescriptionLabels(allocator, descriptions);
    allocator.free(descriptions);
}

/// Parsed `VAL_` payload before it is attached to a signal.
pub const ValueDescriptionRef = union(enum) {
    table_name: []const u8,
    inline_values: []ValueDescription,
};

/// Signal-specific value descriptions from a `VAL_` record.
pub const SignalValueDescriptions = struct {
    message_id: u32,
    signal_name: []const u8,
    value_descriptions: ValueDescriptionRef,

    /// Parses either inline value descriptions or a named value-table reference.
    pub fn fromString(allocator: std.mem.Allocator, line: []const u8) !SignalValueDescriptions {
        var cursor = std.mem.trim(u8, line, " \t\r");
        if (!std.mem.startsWith(u8, cursor, "VAL_ ")) return error.InvalidValueDescriptionLine;
        cursor = std.mem.trim(u8, cursor["VAL_ ".len..], " \t");

        const message_id_end = std.mem.indexOfAny(u8, cursor, " \t") orelse return error.InvalidValueDescriptionLine;
        const message_id = try std.fmt.parseInt(u32, cursor[0..message_id_end], 10);
        cursor = std.mem.trim(u8, cursor[message_id_end..], " \t");

        const signal_name_end = std.mem.indexOfAny(u8, cursor, " \t") orelse return error.InvalidValueDescriptionLine;
        const signal_name = cursor[0..signal_name_end];
        cursor = std.mem.trim(u8, cursor[signal_name_end..], " \t");

        if (cursor.len == 0) return error.InvalidValueDescriptionLine;
        if (std.mem.endsWith(u8, cursor, ";")) {
            cursor = std.mem.trim(u8, cursor[0 .. cursor.len - 1], " \t\r");
        }
        if (cursor.len == 0) return error.InvalidValueDescriptionLine;

        return .{
            .message_id = message_id,
            .signal_name = signal_name,
            .value_descriptions = switch (cursor[0]) {
                '-', '0'...'9' => .{ .inline_values = try parseValueDescriptionPairs(allocator, cursor) },
                '"' => return error.InvalidValueDescriptionLine,
                else => .{ .table_name = cursor },
            },
        };
    }
};

/// Named set of value descriptions from a `VAL_TABLE_` record.
pub const ValueTable = struct {
    name: []const u8,
    values: []ValueDescription,

    /// Parses a named table and its raw-value/label pairs.
    pub fn fromString(allocator: std.mem.Allocator, line: []const u8) !ValueTable {
        var cursor = std.mem.trim(u8, line, " \t\r");
        if (!std.mem.startsWith(u8, cursor, "VAL_TABLE_ ")) return error.InvalidValueTableLine;
        cursor = std.mem.trim(u8, cursor["VAL_TABLE_ ".len..], " \t");

        const name_end = std.mem.indexOfAny(u8, cursor, " \t") orelse return error.InvalidValueTableLine;
        const name = cursor[0..name_end];
        cursor = std.mem.trim(u8, cursor[name_end..], " \t");

        return .{
            .name = name,
            .values = try parseValueDescriptionPairs(allocator, cursor),
        };
    }
};

/// Signal numeric type metadata from a `SIG_VALTYPE_` record.
pub const SignalValueType = struct {
    message_id: u32,
    signal_name: []const u8,
    value_type: ValueType,

    /// Parses the integer type code used by DBC value-type metadata.
    pub fn fromString(line: []const u8) !SignalValueType {
        var cursor = std.mem.trim(u8, line, " \t\r");
        if (!std.mem.startsWith(u8, cursor, "SIG_VALTYPE_ ")) return error.InvalidSignalValueTypeLine;
        cursor = std.mem.trim(u8, cursor["SIG_VALTYPE_ ".len..], " \t");

        const message_id_end = std.mem.indexOfAny(u8, cursor, " \t") orelse return error.InvalidSignalValueTypeLine;
        const message_id = try std.fmt.parseInt(u32, cursor[0..message_id_end], 10);
        cursor = std.mem.trim(u8, cursor[message_id_end..], " \t");

        const signal_name_end = std.mem.indexOfAny(u8, cursor, " \t:") orelse return error.InvalidSignalValueTypeLine;
        const signal_name = cursor[0..signal_name_end];
        cursor = std.mem.trim(u8, cursor[signal_name_end..], " \t");
        if (std.mem.startsWith(u8, cursor, ":")) {
            cursor = std.mem.trim(u8, cursor[1..], " \t");
        }
        if (std.mem.endsWith(u8, cursor, ";")) {
            cursor = std.mem.trim(u8, cursor[0 .. cursor.len - 1], " \t\r");
        }

        return .{
            .message_id = message_id,
            .signal_name = signal_name,
            .value_type = switch (try std.fmt.parseInt(u8, cursor, 10)) {
                0 => .integer,
                1 => .float32,
                2 => .float64,
                else => return error.InvalidSignalValueTypeLine,
            },
        };
    }
};

/// Parses repeated `<raw> "<label>"` pairs.
fn parseValueDescriptionPairs(allocator: std.mem.Allocator, text: []const u8) ![]ValueDescription {
    var cursor = std.mem.trim(u8, text, " \t\r");
    if (std.mem.endsWith(u8, cursor, ";")) {
        cursor = std.mem.trim(u8, cursor[0 .. cursor.len - 1], " \t\r");
    }

    var descriptions: std.ArrayList(ValueDescription) = .empty;
    errdefer {
        freeValueDescriptionLabels(allocator, descriptions.items);
        descriptions.deinit(allocator);
    }

    while (cursor.len > 0) {
        const raw_end = std.mem.indexOfAny(u8, cursor, " \t") orelse return error.InvalidValueDescriptionLine;
        const raw_value = try std.fmt.parseInt(i64, cursor[0..raw_end], 10);
        try ensureJsSafeInteger(raw_value);
        cursor = std.mem.trim(u8, cursor[raw_end..], " \t");

        const label = quotes.parseQuoted(allocator, &cursor) catch |err| switch (err) {
            error.OutOfMemory => return err,
            else => return error.InvalidValueDescriptionLine,
        };
        descriptions.append(allocator, .{
            .raw_value = raw_value,
            .label = label,
        }) catch |err| {
            allocator.free(label);
            return err;
        };
        cursor = std.mem.trim(u8, cursor, " \t");
    }

    return descriptions.toOwnedSlice(allocator);
}

/// Keeps JSON-facing raw values exact when they reach TypeScript.
fn ensureJsSafeInteger(value: i64) !void {
    if (value < -JS_SAFE_INTEGER_MAX or value > JS_SAFE_INTEGER_MAX) {
        return error.RawValueOutsideJsSafeIntegerRange;
    }
}

test "parse fixture VAL line" {
    const allocator = std.testing.allocator;
    const signal_values = try SignalValueDescriptions.fromString(allocator, "VAL_ 100 State 0 \"Off\" 1 \"On\";");
    const descriptions = switch (signal_values.value_descriptions) {
        .inline_values => |items| items,
        .table_name => unreachable,
    };
    defer freeValueDescriptions(allocator, descriptions);

    try std.testing.expectEqual(@as(u32, 100), signal_values.message_id);
    try std.testing.expectEqualStrings("State", signal_values.signal_name);
    try std.testing.expectEqual(@as(usize, 2), descriptions.len);
    try std.testing.expectEqual(@as(i64, 0), descriptions[0].raw_value);
    try std.testing.expectEqualStrings("Off", descriptions[0].label);
    try std.testing.expectEqual(@as(i64, 1), descriptions[1].raw_value);
    try std.testing.expectEqualStrings("On", descriptions[1].label);
}

test "parse VAL line referencing table" {
    const allocator = std.testing.allocator;
    const signal_values = try SignalValueDescriptions.fromString(allocator, "VAL_ 100 State GearStates;");
    const table_name = switch (signal_values.value_descriptions) {
        .inline_values => unreachable,
        .table_name => |name| name,
    };

    try std.testing.expectEqual(@as(u32, 100), signal_values.message_id);
    try std.testing.expectEqualStrings("State", signal_values.signal_name);
    try std.testing.expectEqualStrings("GearStates", table_name);
}

test "parse VAL_TABLE line" {
    const allocator = std.testing.allocator;
    const table = try ValueTable.fromString(allocator, "VAL_TABLE_ GearStates 0 \"Park\" 1 \"Drive\" 2 \"Reverse\";");
    defer freeValueDescriptions(allocator, table.values);

    try std.testing.expectEqualStrings("GearStates", table.name);
    try std.testing.expectEqual(@as(usize, 3), table.values.len);
    try std.testing.expectEqual(@as(i64, 2), table.values[2].raw_value);
    try std.testing.expectEqualStrings("Reverse", table.values[2].label);
}

test "parse signed value descriptions" {
    const allocator = std.testing.allocator;
    const table = try ValueTable.fromString(allocator, "VAL_TABLE_ SignedStates -1 \"Unknown\" 0 \"Off\";");
    defer freeValueDescriptions(allocator, table.values);

    try std.testing.expectEqual(@as(i64, -1), table.values[0].raw_value);
    try std.testing.expectEqualStrings("Unknown", table.values[0].label);
}

test "reject VAL line without VAL prefix" {
    const allocator = std.testing.allocator;
    try std.testing.expectError(error.InvalidValueDescriptionLine, SignalValueDescriptions.fromString(allocator, "VAL_TABLE_ State 0 \"Off\";"));
}

test "reject VAL line with missing label quote" {
    const allocator = std.testing.allocator;
    try std.testing.expectError(error.InvalidValueDescriptionLine, SignalValueDescriptions.fromString(allocator, "VAL_ 100 State 0 Off;"));
}

test "reject VAL line with non-numeric raw value" {
    const allocator = std.testing.allocator;
    try std.testing.expectError(error.InvalidCharacter, SignalValueDescriptions.fromString(allocator, "VAL_ 100 State 0x1 \"Off\";"));
}

test "reject VAL line with unsafe raw value" {
    const allocator = std.testing.allocator;
    try std.testing.expectError(error.RawValueOutsideJsSafeIntegerRange, SignalValueDescriptions.fromString(allocator, "VAL_ 100 State 9007199254740992 \"Too Big\";"));
}

test "parse VAL line with escaped label" {
    const allocator = std.testing.allocator;
    const signal_values = try SignalValueDescriptions.fromString(allocator, "VAL_ 100 State 1 \"State \\\"On\\\"\";");
    const descriptions = switch (signal_values.value_descriptions) {
        .inline_values => |items| items,
        .table_name => unreachable,
    };
    defer freeValueDescriptions(allocator, descriptions);

    try std.testing.expectEqualStrings("State \"On\"", descriptions[0].label);
}

test "reject VAL_TABLE line without table name" {
    const allocator = std.testing.allocator;
    try std.testing.expectError(error.InvalidValueTableLine, ValueTable.fromString(allocator, "VAL_TABLE_"));
}

test "reject VAL_TABLE line with unterminated label" {
    const allocator = std.testing.allocator;
    try std.testing.expectError(error.InvalidValueDescriptionLine, ValueTable.fromString(allocator, "VAL_TABLE_ State 0 \"Off;"));
}

test "parse SIG_VALTYPE line with colon" {
    const value_type = try SignalValueType.fromString("SIG_VALTYPE_ 100 Temperature : 1;");

    try std.testing.expectEqual(@as(u32, 100), value_type.message_id);
    try std.testing.expectEqualStrings("Temperature", value_type.signal_name);
    try std.testing.expectEqual(ValueType.float32, value_type.value_type);
}

test "parse SIG_VALTYPE line without colon" {
    const value_type = try SignalValueType.fromString("SIG_VALTYPE_ 100 Temperature 2;");

    try std.testing.expectEqual(ValueType.float64, value_type.value_type);
}

test "reject SIG_VALTYPE line with unsupported type" {
    try std.testing.expectError(error.InvalidSignalValueTypeLine, SignalValueType.fromString("SIG_VALTYPE_ 100 Temperature : 3;"));
}

File: wasm/src/dbc/values.rs
use std::rc::Rc;

use super::quotes::parse_quoted;
use super::{DbcError, find_dbc_whitespace, trim_dbc, trim_space_tab};

/// Largest exact integer representable by JavaScript `number`.
const JS_SAFE_INTEGER_MAX: i64 = 9_007_199_254_740_991;

/// Numeric representation requested by `SIG_VALTYPE_`.
#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub enum ValueType {
    #[default]
    Integer,
    Float32,
    Float64,
}

impl ValueType {
    pub(crate) const fn as_catalog_str(self) -> &'static str {
        match self {
            Self::Integer => "integer",
            Self::Float32 => "float32",
            Self::Float64 => "float64",
        }
    }
}

/// One raw numeric value and its display label.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ValueDescription {
    pub raw_value: i64,
    pub label: String,
}

/// Named set of value descriptions from a `VAL_TABLE_` record.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ValueTable {
    pub name: String,
    pub values: Rc<[ValueDescription]>,
}

impl ValueTable {
    /// Parses a named table and its raw-value/label pairs.
    pub fn parse(line: &str) -> Result<Self, DbcError> {
        let mut cursor = trim_dbc(line);
        let Some(rest) = cursor.strip_prefix("VAL_TABLE_ ") else {
            return Err(DbcError::InvalidValueTableLine);
        };
        cursor = trim_space_tab(rest);

        let Some(name_end) = find_dbc_whitespace(cursor) else {
            return Err(DbcError::InvalidValueTableLine);
        };
        let name = cursor[..name_end].to_owned();
        cursor = trim_space_tab(&cursor[name_end..]);

        Ok(Self {
            name,
            values: parse_value_description_pairs(cursor)?.into(),
        })
    }
}

/// A `VAL_` record's value source before it is attached to a signal.
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum ValueDescriptionRef {
    TableName(String),
    InlineValues(Rc<[ValueDescription]>),
}

/// Signal-specific value descriptions from a `VAL_` record.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SignalValueDescriptions {
    pub message_id: u32,
    pub signal_name: String,
    pub value_descriptions: ValueDescriptionRef,
}

impl SignalValueDescriptions {
    /// Parses either inline descriptions or a named value-table reference.
    pub fn parse(line: &str) -> Result<Self, DbcError> {
        let mut cursor = trim_dbc(line);
        let Some(rest) = cursor.strip_prefix("VAL_ ") else {
            return Err(DbcError::InvalidValueDescriptionLine);
        };
        cursor = trim_space_tab(rest);

        let Some(message_id_end) = find_dbc_whitespace(cursor) else {
            return Err(DbcError::InvalidValueDescriptionLine);
        };
        let message_id_text = &cursor[..message_id_end];
        let message_id = message_id_text.parse().map_err(|error| {
            DbcError::invalid_integer("value-description message ID", message_id_text, error)
        })?;
        cursor = trim_space_tab(&cursor[message_id_end..]);

        let Some(signal_name_end) = find_dbc_whitespace(cursor) else {
            return Err(DbcError::InvalidValueDescriptionLine);
        };
        let signal_name = cursor[..signal_name_end].to_owned();
        cursor = trim_space_tab(&cursor[signal_name_end..]);

        if cursor.is_empty() {
            return Err(DbcError::InvalidValueDescriptionLine);
        }
        cursor = strip_record_semicolon(cursor);
        if cursor.is_empty() {
            return Err(DbcError::InvalidValueDescriptionLine);
        }

        let value_descriptions = match cursor.as_bytes()[0] {
            b'-' | b'0'..=b'9' => {
                ValueDescriptionRef::InlineValues(parse_value_description_pairs(cursor)?.into())
            }
            b'"' => return Err(DbcError::InvalidValueDescriptionLine),
            _ => ValueDescriptionRef::TableName(cursor.to_owned()),
        };

        Ok(Self {
            message_id,
            signal_name,
            value_descriptions,
        })
    }
}

/// Signal numeric type metadata from a `SIG_VALTYPE_` record.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SignalValueType {
    pub message_id: u32,
    pub signal_name: String,
    pub value_type: ValueType,
}

impl SignalValueType {
    /// Parses the integer type code used by DBC value-type metadata.
    pub fn parse(line: &str) -> Result<Self, DbcError> {
        let mut cursor = trim_dbc(line);
        let Some(rest) = cursor.strip_prefix("SIG_VALTYPE_ ") else {
            return Err(DbcError::InvalidSignalValueTypeLine);
        };
        cursor = trim_space_tab(rest);

        let Some(message_id_end) = find_dbc_whitespace(cursor) else {
            return Err(DbcError::InvalidSignalValueTypeLine);
        };
        let message_id_text = &cursor[..message_id_end];
        let message_id = message_id_text.parse().map_err(|error| {
            DbcError::invalid_integer("signal value-type message ID", message_id_text, error)
        })?;
        cursor = trim_space_tab(&cursor[message_id_end..]);

        let Some(signal_name_end) = cursor
            .as_bytes()
            .iter()
            .position(|byte| matches!(byte, b' ' | b'\t' | b'\r' | b':'))
        else {
            return Err(DbcError::InvalidSignalValueTypeLine);
        };
        let signal_name = cursor[..signal_name_end].to_owned();
        cursor = trim_space_tab(&cursor[signal_name_end..]);
        if let Some(rest) = cursor.strip_prefix(':') {
            cursor = trim_space_tab(rest);
        }
        cursor = strip_record_semicolon(cursor);

        let value_type_code: u8 = cursor
            .parse()
            .map_err(|error| DbcError::invalid_integer("signal value type", cursor, error))?;
        let value_type = match value_type_code {
            0 => ValueType::Integer,
            1 => ValueType::Float32,
            2 => ValueType::Float64,
            _ => return Err(DbcError::InvalidSignalValueTypeLine),
        };

        Ok(Self {
            message_id,
            signal_name,
            value_type,
        })
    }
}

fn strip_record_semicolon(cursor: &str) -> &str {
    cursor
        .strip_suffix(';')
        .map_or(cursor, |without_semicolon| trim_dbc(without_semicolon))
}

/// Parses repeated `<raw> "<label>"` pairs.
fn parse_value_description_pairs(text: &str) -> Result<Vec<ValueDescription>, DbcError> {
    let mut cursor = strip_record_semicolon(trim_dbc(text));
    let mut descriptions = Vec::new();

    while !cursor.is_empty() {
        let Some(raw_end) = find_dbc_whitespace(cursor) else {
            return Err(DbcError::InvalidValueDescriptionLine);
        };
        let raw_text = &cursor[..raw_end];
        let raw_value: i64 = raw_text.parse().map_err(|error| {
            DbcError::invalid_integer("value-description raw value", raw_text, error)
        })?;
        ensure_js_safe_integer(raw_value)?;
        cursor = trim_space_tab(&cursor[raw_end..]);

        let (label, rest) = parse_quoted(cursor).map_err(|error| match error {
            DbcError::InvalidQuotedString => DbcError::InvalidValueDescriptionLine,
            other => other,
        })?;
        descriptions.push(ValueDescription { raw_value, label });
        cursor = trim_space_tab(rest);
    }

    Ok(descriptions)
}

fn ensure_js_safe_integer(value: i64) -> Result<(), DbcError> {
    if !(-JS_SAFE_INTEGER_MAX..=JS_SAFE_INTEGER_MAX).contains(&value) {
        return Err(DbcError::RawValueOutsideJsSafeIntegerRange(value));
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_inline_value_descriptions() {
        let signal_values =
            SignalValueDescriptions::parse("VAL_ 100 State 0 \"Off\" 1 \"On\";").unwrap();
        let ValueDescriptionRef::InlineValues(descriptions) = signal_values.value_descriptions
        else {
            panic!("expected inline values");
        };

        assert_eq!(signal_values.message_id, 100);
        assert_eq!(signal_values.signal_name, "State");
        assert_eq!(descriptions.len(), 2);
        assert_eq!(descriptions[0].raw_value, 0);
        assert_eq!(descriptions[0].label, "Off");
        assert_eq!(descriptions[1].raw_value, 1);
        assert_eq!(descriptions[1].label, "On");
    }

    #[test]
    fn parses_value_table_reference() {
        let signal_values = SignalValueDescriptions::parse("VAL_ 100 State GearStates;").unwrap();

        assert_eq!(signal_values.message_id, 100);
        assert_eq!(signal_values.signal_name, "State");
        assert!(matches!(
            signal_values.value_descriptions,
            ValueDescriptionRef::TableName(ref name) if name == "GearStates"
        ));
    }

    #[test]
    fn parses_value_table() {
        let table =
            ValueTable::parse("VAL_TABLE_ GearStates 0 \"Park\" 1 \"Drive\" 2 \"Reverse\";")
                .unwrap();

        assert_eq!(table.name, "GearStates");
        assert_eq!(table.values.len(), 3);
        assert_eq!(table.values[2].raw_value, 2);
        assert_eq!(table.values[2].label, "Reverse");
    }

    #[test]
    fn parses_signed_value_descriptions() {
        let table = ValueTable::parse("VAL_TABLE_ SignedStates -1 \"Unknown\" 0 \"Off\";").unwrap();

        assert_eq!(table.values[0].raw_value, -1);
        assert_eq!(table.values[0].label, "Unknown");
    }

    #[test]
    fn rejects_wrong_value_description_prefix() {
        assert!(matches!(
            SignalValueDescriptions::parse("VAL_TABLE_ State 0 \"Off\";"),
            Err(DbcError::InvalidValueDescriptionLine)
        ));
    }

    #[test]
    fn rejects_missing_value_label_quote() {
        assert!(matches!(
            SignalValueDescriptions::parse("VAL_ 100 State 0 Off;"),
            Err(DbcError::InvalidValueDescriptionLine)
        ));
    }

    #[test]
    fn rejects_non_numeric_raw_value() {
        assert!(matches!(
            SignalValueDescriptions::parse("VAL_ 100 State 0x1 \"Off\";"),
            Err(DbcError::InvalidInteger { .. })
        ));
    }

    #[test]
    fn rejects_raw_value_outside_javascript_safe_range() {
        assert!(matches!(
            SignalValueDescriptions::parse("VAL_ 100 State 9007199254740992 \"Too Big\";"),
            Err(DbcError::RawValueOutsideJsSafeIntegerRange(
                9_007_199_254_740_992
            ))
        ));
    }

    #[test]
    fn parses_escaped_value_label() {
        let signal_values =
            SignalValueDescriptions::parse("VAL_ 100 State 1 \"State \\\"On\\\"\";").unwrap();
        let ValueDescriptionRef::InlineValues(descriptions) = signal_values.value_descriptions
        else {
            panic!("expected inline values");
        };

        assert_eq!(descriptions[0].label, "State \"On\"");
    }

    #[test]
    fn rejects_value_table_without_name() {
        assert!(matches!(
            ValueTable::parse("VAL_TABLE_"),
            Err(DbcError::InvalidValueTableLine)
        ));
    }

    #[test]
    fn rejects_unterminated_table_label() {
        assert!(matches!(
            ValueTable::parse("VAL_TABLE_ State 0 \"Off;"),
            Err(DbcError::InvalidValueDescriptionLine)
        ));
    }

    #[test]
    fn parses_signal_value_type_with_colon() {
        let value_type = SignalValueType::parse("SIG_VALTYPE_ 100 Temperature : 1;").unwrap();

        assert_eq!(value_type.message_id, 100);
        assert_eq!(value_type.signal_name, "Temperature");
        assert_eq!(value_type.value_type, ValueType::Float32);
    }

    #[test]
    fn parses_signal_value_type_without_colon() {
        let value_type = SignalValueType::parse("SIG_VALTYPE_ 100 Temperature 2;").unwrap();

        assert_eq!(value_type.value_type, ValueType::Float64);
    }

    #[test]
    fn rejects_unsupported_signal_value_type() {
        assert!(matches!(
            SignalValueType::parse("SIG_VALTYPE_ 100 Temperature : 3;"),
            Err(DbcError::InvalidSignalValueTypeLine)
        ));
    }
}
